### ISSUE TITLE
read.rsk() should be able to read location data

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#2024]
A user, who is not yet on github but who I hope will join and contribute to this thread, asked whether `read.rsk()` could read location data.  I am going to add that, in a new branch called `rsk` that hangs off the `develop` branch.

I have renamed the file the user sent as `ctd_with_location.rsk` (to protect privacy) and will be making notes here on what I learn about the file.  Hopefully, in a few days, the user will join github and can give me and Clark pointers, if we need them.

**Comment by dankelley** [comment]
Hm. As below, I tried examining the file in R and don't quite see which table might control longitude and latitude.

Next, I'll use a GUI-based sqlite viewer, which makes it easier to drill down.
```R
> f <- "~/ctd_with_location.rsk"
> library(DBI)
> con <- DBI::dbConnect(RSQLite::SQLite(), f)
> sort(dbListTables(con))
 [1] "android_metadata"      "average"               "breaksdata"           
 [4] "burst"                 "caldata"               "calsource"            
 [7] "channels"              "chartView"             "continuous"           
[10] "data"                  "datasetDeployments"    "datasets"             
[13] "dbInfo"                "deployments"           "directional"          
[16] "downloadprogress"      "downloads"             "downloadsevents"      
[19] "downloadsheader"       "downsample_caches"     "epochs"               
[22] "errors"                "events"                "geodata"              
[25] "instrumentChannels"    "instruments"           "parameters"           
[28] "power"                 "regimes"               "region"               
[31] "regionCal"             "regionCast"            "regionComment"        
[34] "regionExclude"         "regionGeoData"         "regionPlateau"        
[37] "regionProfile"         "schedules"             "sqlite_sequence"      
[40] "thumbnailData"         "tide"                  "valve"                
[43] "valveSchedule"         "valveScheduleSegments" "wave"                 
[46] "wifi"              
```

**Comment by dankelley** [comment]
OK, using `DB Browser for SQLite` on my mac suggests looking in `region` and I see as follows, which looks promising.  Next I'll see if the row count matches the row count of e.g. pressure.

(Note: I am not showing the actual data here because I don't know whether they are private, and I start with that assumption unless an issue-reporter says otherwise. Besides, I don't see a reason -- yet -- to try sample plots.)

```
> geodata <- dbReadTable(con, "geodata")
 names(geodata)
[1] "tstamp"       "latitude"     "longitude"    "accuracy"     "accuracyType"
[6] "origin"  
```

**Comment by dankelley** [comment]
I have verified that this `geodata` field is not present in the `~/git/oce/tests/testthat/local_data/060130_20150904_1159.rsk` file that is present in the oce repo.

**Comment by dankelley** [comment]
@richardsc -- I will likely need advice on whether lon and lat are in the whacky order that other fields might be in.  (To the original reporter: RSK files, or at least some of them, do not hold data in time-increasing order.)

For now, I'm going to order lon and lat as the other data are ordered.  That can be changed later. The relevant bit of code is as below (line numbers are not as in GH since I am not going to push any of this until I have more things working).

Frankly, dealing with the `from`, `to` and `by` arguments is going to be harder than just reading the lon and lat data.  That's just the way it is, with these arguments, in many oce functions.

<img width="767" alt="Screenshot 2022-12-10 at 8 51 05 AM" src="https://user-images.githubusercontent.com/99469/206856141-5c4762ef-a4d3-4d54-bd2c-978723c3c262.png">

**Comment by dankelley** [comment]
Uh oh.  The length of the `longitude` and `latitude` vectors does not match the length of the `time` vector, as shown below for the sample file from the user.  (This code not yet pushed to GH.)

```R
> source("R/rsk.R");d<-read.rsk(f,debug=1)
      read.rsk(file="~/ctd_with_location.rsk", from=1, to=(not given), by=1, type=(missing), tz="UTC", ...) {
        this file contains 15491 longitude and latitude data
        first, guess pressureAtmospheric=10.1325
        after studying the RSK file, now have pressureAtmospheric=10.1325
        SQL query to fetch data: "SELECT tstamp,channel01,channel02,channel03,channel04,channel05,channel06,channel07,channel08,channel09 FROM data ;"
      } # read.rsk()
Warning message:
In read.rsk(f, debug = 1) :
  the length of the 'time' vector (17396) does not match the 'longitude' and 'latitude' lengths (15491) so location cannot be reordered for time
```

**Comment by dankelley** [comment]
Since `geodata` has a field named `tstamp`, maybe a way to reconcile the different lengths of the location data and the ctd data is to key by that value (directly or using `approx()`, perhaps).

**Comment by dankelley** [comment]
In commit a74a0e497eaabf1d86c383dc0fb9a37c4bf924eb of the `rsk` branch, I've made some changes to tackle this issue.  I will leave this alone for a while, to think about the best way to merge the geodata and the data (as discussed above, and in private emails).

(To the person who started this, over email -- please click the link in the previous paragraph to see the changes, and a detailed commit note.)

@richardsc -- I could do a zoom call next week to discuss this.  I won't be in the office much, with exams and such.

**Comment by dankelley** [comment]
I get as follows.  Note the roughly 8-hour shift, which might be a result of using local time for one stream and UTC for another (given the location of the dataset).
```R
> library(oce)
> d<-read.oce("~/ctd_with_location.rsk")
> tg<-numberAsPOSIXct(as.numeric(d[["geodata"]]$tstamp)/1e3,type="unix")
> range(d[["time"]])
[1] "2022-08-08 15:09:43.0 UTC" "2022-08-10 16:00:34.5 UTC"
> range(tg)
[1] "2022-08-08 23:04:58.079 UTC" "2022-08-11 00:00:44.647 UTC"
```

**Comment by dankelley** [comment]
I looked a bit in the files and think I see how to navigate the data to find downcasts and upcasts, but doing so involves reading some tables from the SQL file that have not been read by `read.rsk()` so far.

My guess is that we look in the `regionCast` table, isolate cases where `type=="DOWN"`, and grab the `regionID` column from that.  Then we match that to the `regionID` column of the `region` table, and extract the `tstamp1` and `tstamp2` values.  Finally, we use those index pairs to look within the `data` slot, which now stores the `tstamp` column, for this purpose.  I'm not sure why I just wrote this paragraph, because the code would be shorter, but I suppose I'd like to get this "on paper" now, for further discussion, before coding.  There may be issues with the fact that the `tstamp` column is kind of weird, being of type `int64`, which will not work on 32-bit machines ... but frankly I don't see any reason for users to be denied features because they are working on an archaic machine.

Therefore, I added a new arg to `read.rsk()` called `allTables`, which will cause every table to be read and stored in `metadata`.  That way, a user who understands the structure can do whatever they want with the data, e.g. isolating casts, working with location data, etc. At the moment, `allTables` is defaulting to FALSE, but changing that is trivial, of course.

Below is a demo.  (PS: 

<details>

```R
> library(oce)
> f <- "ctd_with_location.rsk"
> d <- read.rsk(f, allTables=TRUE)
Warning messages:
1: In read.rsk(f, allTables = TRUE) :
  FIXME: provide a way to merge metadata$geodata with data
2: In read.rsk(f, allTables = TRUE) :
  to convert RSK timestamp to time: numberAsPOSIXct(as.numeric(tstamp)/1e3,type="unix")
> str(d@data,1)
List of 6
 $ time        : POSIXct[1:17396], format: "2022-08-08 15:09:43.0" ...
 $ conductivity: num [1:17396] -1.63e-04 -3.91e-04 -3.34e-04 -1.78e-04 -2.93e-05 ...
 $ temperature : num [1:17396] 9.38 9.37 9.38 9.4 9.41 ...
 $ pressure    : num [1:17396] 10.1 10.1 10.1 10.1 10.1 ...
 $ turbidity   : num [1:17396] 0.349 0.445 1.85 0.264 1.759 ...
 $ tstamp      :integer64 [1:17396] 1659971383000 1659971383500 1659971384000 1659971384500 1659971385000 1659971385500 1659971386000 1659971386500 ... 
> str(d@metadata,1)
List of 57
 $ units                :List of 4
 $ flags                : list()
 $ filename             : chr "ctd_with_location.rsk"
 $ model                : chr "RBRconcerto3"
 $ pressureType         : chr "absolute"
 $ pressureAtmospheric  : num 10.1
 $ android_metadata     :'data.frame':	1 obs. of  1 variable:
 $ average              :'data.frame':	0 obs. of  5 variables:
 $ breaksdata           :'data.frame':	28 obs. of  2 variables:
 $ burst                :'data.frame':	0 obs. of  5 variables:
 $ caldata              :'data.frame':	0 obs. of  3 variables:
 $ calsource            :'data.frame':	0 obs. of  2 variables:
 $ channels             :'data.frame':	9 obs. of  9 variables:
 $ chartView            :'data.frame':	15 obs. of  3 variables:
 $ continuous           :'data.frame':	0 obs. of  3 variables:
 $ data                 :'data.frame':	17396 obs. of  10 variables:
 $ datasetDeployments   :'data.frame':	0 obs. of  2 variables:
 $ datasets             :'data.frame':	0 obs. of  2 variables:
 $ dbInfo               :'data.frame':	2 obs. of  2 variables:
 $ deployments          :'data.frame':	1 obs. of  10 variables:
 $ directional          :'data.frame':	0 obs. of  7 variables:
 $ downloadprogress     :'data.frame':	1 obs. of  3 variables:
 $ downloads            :'data.frame':	2215 obs. of  4 variables:
 $ downloadsevents      :'data.frame':	63 obs. of  4 variables:
 $ downloadsheader      :'data.frame':	1 obs. of  4 variables:
 $ downsample_caches    :'data.frame':	1 obs. of  2 variables:
 $ epochs               :'data.frame':	1 obs. of  4 variables:
 $ errors               :'data.frame':	0 obs. of  5 variables:
 $ events               :'data.frame':	204 obs. of  5 variables:
 $ geodata              :'data.frame':	15491 obs. of  6 variables:
 $ instrumentChannels   :'data.frame':	9 obs. of  4 variables:
 $ instruments          :'data.frame':	1 obs. of  2 variables:
 $ parameters           :'data.frame':	1 obs. of  2 variables:
 $ power                :'data.frame':	0 obs. of  13 variables:
 $ regimes              :'data.frame':	0 obs. of  5 variables:
 $ region               :'data.frame':	111 obs. of  8 variables:
 $ regionCal            :'data.frame':	0 obs. of  4 variables:
 $ regionCast           :'data.frame':	46 obs. of  3 variables:
 $ regionComment        :'data.frame':	0 obs. of  1 variable:
 $ regionExclude        :'data.frame':	0 obs. of  3 variables:
 $ regionGeoData        :'data.frame':	42 obs. of  3 variables:
 $ regionPlateau        :'data.frame':	0 obs. of  4 variables:
 $ regionProfile        :'data.frame':	23 obs. of  1 variable:
 $ schedules            :'data.frame':	1 obs. of  8 variables:
 $ sqlite_sequence      :'data.frame':	0 obs. of  2 variables:
 $ thumbnailData        :'data.frame':	0 obs. of  5 variables:
 $ tide                 :'data.frame':	0 obs. of  5 variables:
 $ valve                :'data.frame':	0 obs. of  10 variables:
 $ valveSchedule        :'data.frame':	0 obs. of  7 variables:
 $ valveScheduleSegments:'data.frame':	0 obs. of  5 variables:
 $ wave                 :'data.frame':	0 obs. of  5 variables:
 $ wifi                 :'data.frame':	0 obs. of  3 variables:
 $ dataNamesOriginal    : Named chr [1:6] "-" "cond19" "temp14" "pres24" ...
  ..- attr(*, "names")= chr [1:6] "time" "conductivity" "temperature" "pressure" ...
 $ serialNumber         : int 209642
 $ sampleInterval       : num 0.5
 $ rskVersion           : num [1:3] 1 3 0
 $ ruskinVersion        : chr "mobile"
```

**Comment by richardsc** [comment]
I think the above should be it's own, granular issue (so as not to confuse things here, which is dealing with location data), and have split it off as such.

**Comment by dankelley** [comment]
NOTE: the split-off issues is #2025

**Comment by dankelley** [comment]
This issue has been addressed (as discussed here and, I admit confusingly, in several other issues) so I am closing it.

The "rsk" branch has been merged into the "develop" branch (commit c3bb3fd3e517cc2cd94fbd288ca1767a2a1ec839) and it is the latter that ought to be used for further tests.  

If large-scale recoding is required to address any issues, I'll revive "rsk" but for now, I ask @Brendan-Higgins and @richardsc (the two most involved in the RSK work over the past few weeks) to focus on "develop".  Also, I ask that they create new issues with narrow titles, if new problems arise or new features are desired.  Issue drift gets confusing on github, owing to the lack of threaded comments.

Thanks for the patience on this.

Dan.

**Comment by dankelley** [comment]
I'm reopening this, because one of my testing suites now reveals a problem.  I'll address this tomorrow morning.
```R

R version 4.2.2 (2022-10-31) -- "Innocent and Trusting"
Copyright (C) 2022 The R Foundation for Statistical Computing
Platform: x86_64-apple-darwin17.0 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> library(oce)
Loading required package: gsw
> options(width=100)                     # makes summaries easier to read
> file <- "~/Dropbox/oce_secret_data/01.rsk"
> if (file.exists(file)) {
+     d <- read.oce(file)
+     summary(d)
+     if (!interactive()) png("rsk01a.png")
+     plot(d)
+     if (!interactive()) dev.off()
+     if (!interactive()) png("rsk01b.png")
+     par(mfrow=c(2, 1))
+     hist(d[["cond12"]])
+     hist(d[["conductivity"]])
+     if (!interactive()) dev.off()
+ }
Error in approx(geodata$time[look], geodata$latitude[look], res@data$time,  : 
  need at least two non-NA values to interpolate
Calls: read.oce -> read.rsk -> approx
Execution halted

```

**Comment by dankelley** [comment]
Hm.  I see the problem with this older file (I cannot say more about it because I don't know if it can be shared publicly): the code that I wrote at the start of this thread assumes that the `geodata` table in the `.rsk` file contains something called `origin`, and it limits its focus to cases where that equals `"auto"`, viz.
https://github.com/dankelley/oce/blob/da7a7db71a35f77e8963ec6df617fdf9b65084a2/R/rsk.R#L1014

so what I'll do next is to try restricting only if `origin` exists.

**Comment by dankelley** [comment]
This problem with the old file is fixed in "develop" commit 84d11301ce956c49766462379068296e664118da.  The whole test/check suite now passes cleanly, so I am (re)closing this issue.
